README.txt
=======================
Version 0.1: 2011-09-17
=======================

Base functionality.

=======================
Version 0.2: 2011-09-18
=======================

Ability to delete SMS after reporting them.

Logging cleanups.

Minor typo fix for time reported.

=======================
Version 0.3: 2011-09-21
=======================

Menu.

User can set preferences for automatic delete of reported SMSes, number to SMS violations to.


ChangeLog
2011-09-18  raju  <raju@kandalaya.org>

    * @@@ src/kandalaya/dncviolation/DNCViolationActivity
    
    * Remove unwanted imports.
    
    * Change all logging to debug and set tag to "DNCViolation".
    
    * method doitDoReportButtonHandler: fix typo in time formatter.
    Add functionality for deleting SMS after reporting it. 
    
    * method setSelected: Only update counts if state actually changed.
    
    * @@@ src/kandalaya/dncviolation/DNCViolationDB
    
    * Remove unwanted imports.
    
    * Change all logging to debug and set tag to "DNCViolation".
    
    * Move URI(s) to static members.
    
    * Add methods: deleteSMS(Activity, String) and deleteSMS(Activity, int).
    
    * @@@ AndroidManifest.xml
    
    * Add permission WRITE_SMS.
    
    * Change minSdkVersion (whatever).
    
    * @@@ res/layout/doit.xml
    
    * Add padding to summary.
    
2011-09-20  raju  <raju@kandalaya.org>

    * @@@ src/kandalaya/dncviolation/DNCViolationActivity
    
    * method mainReportButtonHandler: Set default state of delete SMS checkbox from preferences.
    
    * method doitDoReportButtonHandler: Set default destination from preferences.
    
    * Add methods onCreateOptionsMenu, onOptionsItemSelected, showAboutBox.

gen/kandalaya/dncviolation/R.java
/* AUTO-GENERATED FILE.  DO NOT MODIFY.
 *
 * This class was automatically generated by the
 * aapt tool from the resource data it found.  It
 * should not be modified by hand.
 */

package kandalaya.dncviolation;

public final class R {
    public static final class attr {
    }
    public static final class drawable {
        public static final int icon=0x7f020000;
    }
    public static final class id {
        public static final int DATE_CELL=0x7f060016;
        public static final int ItemLayout=0x7f06000b;
        public static final int NAME_CELL=0x7f060015;
        public static final int NUMBER_CELL=0x7f060013;
        public static final int SMSItemLayout=0x7f06000c;
        public static final int actionRow=0x7f060008;
        public static final int callIcon=0x7f06000f;
        public static final int doit=0x7f060000;
        public static final int doit_buttons=0x7f060003;
        public static final int doit_cancel=0x7f060005;
        public static final int doit_delete_sms=0x7f060002;
        public static final int doit_do_report=0x7f060004;
        public static final int doit_summary=0x7f060001;
        public static final int doneButton=0x7f06001f;
        public static final int itemIndex=0x7f06000d;
        public static final int linearLayout1=0x7f060018;
        public static final int linearLayout2=0x7f06001b;
        public static final int listRows=0x7f060007;
        public static final int main=0x7f060006;
        public static final int main_report_button=0x7f060009;
        public static final int numberLayout=0x7f060011;
        public static final int selectViolation=0x7f06000e;
        public static final int selectedCounts=0x7f06000a;
        public static final int smsDate=0x7f06001d;
        public static final int smsDateLabel=0x7f06001c;
        public static final int smsInfo=0x7f060017;
        public static final int smsSender=0x7f06001a;
        public static final int smsSenderLabel=0x7f060019;
        public static final int smsText=0x7f06001e;
        public static final int smsViewButton=0x7f060010;
        public static final int subDetails=0x7f060014;
        public static final int violationDetails=0x7f060012;
    }
    public static final class integer {
        public static final int default_number_of_days_to_go_back=0x7f040000;
    }
    public static final class layout {
        public static final int doit=0x7f030000;
        public static final int main=0x7f030001;
        public static final int row=0x7f030002;
        public static final int smsinfo=0x7f030003;
    }
    public static final class string {
        public static final int app_name=0x7f050002;
        public static final int button_view=0x7f050004;
        public static final int default_reporting_sms_number=0x7f050000;
        public static final int doit_cancel=0x7f05000a;
        public static final int doit_delete_sms=0x7f050008;
        public static final int doit_do_report=0x7f050009;
        public static final int hello=0x7f050001;
        public static final int main_report=0x7f05000b;
        public static final int no_data=0x7f050003;
        public static final int sms_date=0x7f050006;
        public static final int sms_done=0x7f050007;
        public static final int sms_sender=0x7f050005;
    }
}


src/kandalaya/dncviolation/LogItem.java
/*
 * Individual call/SMS log items for reporting DNC violations on Android.
 * 
 * Copyright (C) 2011, Raj Mathur <raju@kandalaya.org>
 * 
 * This file is part of DNCViolation.
 * 
 * DNCViolation is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DNCViolation is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with DNCViolation.  If not, see <http://www.gnu.org/licenses/>.
 */
// %Id$

package kandalaya.dncviolation;

//import kandalaya.dncviolation.DNCViolationID;

/**
 * A single item in the call/SMS log.  Contains the base data from the phone plus additional state and type information.
 * 
 * @author raju
 *
 */
public class LogItem {
	/**
	 * Phone number from which the call/SMS originated.
	 */
	private String
		number = null;
	/**
	 * Cached name of the caller/sender.
	 */
	private String
		name = null;
	/**
	 * Timestamp of the call/SMS in milliseconds since the epoch.
	 */
	private long
		timestamp = -1;
	/**
	 * Whether selected or not in this run.
	 */
	private boolean
		selected;
	/**
	 * Type of record ((C)all or (S)MS)
	 */
	private char
		call_type;
	//private DNCViolationID
	//  id;
	private String
	  id;
	private int
	  index;
	//
	// All we need are getters and setters
	//
	public String getNumber() {
		return number;
	}
	public void setNumber(String number) {
		this.number = number;
	}
	public String getName() {
		return name;
	}
	public void setName(String name) {
		this.name = name;
	}
	public long getTimestamp() {
		return timestamp;
	}
	public void setTimestamp(long timestamp) {
		this.timestamp = timestamp;
	}
	public boolean isSelected() {
		return selected;
	}
	public void setSelected(boolean selected) {
		this.selected = selected;
	}
	public char getCall_type() {
		return call_type;
	}
  public boolean isCall()
  {
    return( call_type == 'C' );
  }
  public boolean isSMS()
  {
    return( call_type == 'S' );
  }
	public void setCall_type(char callType) {
		call_type = callType;
	}
  /**
   * Set the ID based on the number, the name if the number is null, and the timestamp.
   */
	public void setId(String logType, int id)
	{
	  //this.id = new DNCViolationID(makeId(number, name, timestamp));
	  this.id = logType + "-" + id;
	}
	/**
	 * Get the ID of this log item as a string.
	 * @return The hex representation of the ID.
	 */
	public String getId()
	{
	  //return( id.toString() );
	  return( id );
	}
	/**
   * @return the index
   */
  public int getIndex()
  {
    return index;
  }
  /**
   * @param index the index to set
   */
  public void setIndex(int index)
  {
    this.index = index;
  }
  /**
	 * Make a ID hex string out of the given parameters.
	 * @param number The number from which the call/SMS originated.
	 * @param name The cached name of the caller.
	 * @param timestamp Timestamp of the call/SMS.
	 * @return The unique ID of this call/SMS.
	 */
	//public static String makeId( String number, String name, long timestamp )
	//{
	//  return( new DNCViolationID((number == null ?number :name) + timestamp).toString() );
	//}
}


src/kandalaya/dncviolation/DNCViolationID.java
/*
 * Handle SHA-1 ID management for reporting DNC violations on Android.
 * NOT USED!
 * 
 * Copyright (C) 2011, Raj Mathur <raju@kandalaya.org>
 * 
 * This file is part of DNCViolation.
 * 
 * DNCViolation is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DNCViolation is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with DNCViolation.  If not, see <http://www.gnu.org/licenses/>.
 */
package kandalaya.dncviolation;

import java.security.*;

import android.util.Log;

public class DNCViolationID
{
  /**
   * Create an SHA-1 message digest of the given text.
   * 
   * Shamelessly stolen from:
   * http://blog.ryanrampersad.com/2009/02/26/hex-sha1-hashes-in-java/
   * 
   * @param text String to be digested
   * @return byte[] array containing digest
   */
  private static byte[] createHash(String text)
  {
    try
    {
        byte[] b = text.getBytes();
        MessageDigest
          algorithm = MessageDigest.getInstance("SHA-1");
        algorithm.reset();
        algorithm.update(b);
        byte messageDigest[] = algorithm.digest();
        return( messageDigest );
    }
    catch(NoSuchAlgorithmException nsae)
    {
        Log.d("DNCViolation", "no such algorithm exception");
        return null;
    }
  }
  /**
   * Convert the given byte[] array to a hex string.
   * 
   * Shamelessly stolen from:
   * http://blog.ryanrampersad.com/2009/02/26/hex-sha1-hashes-in-java/
   * 
   * @param b array containing values to be converted
   * @return String containing hex representation
   */
  private static String asHex(byte[] b)
  {
    String
      result = "";
    for( int i = 0; i < b.length; i++)
    {
        result += Integer.toString(( b[i] & 0xff ) + 0x100, 16).substring(1);
    }
    return( result );
  }
  /**
   * String from which digest was made.
   */
  String
    text;
  /**
   * Hash of the text.
   */
  byte[]
    hash;
  /**
   * Hex format if user has ever asked for it.
   */
  String
    hex;
  /**
   * Construct a hash ID of the given string.  Store string for later comparisons.
   * @param text string to be digested into an ID.
   */
  public DNCViolationID( String text )
  {
    this.text = text;
    this.hash = createHash(text);
    this.hex = asHex(hash);
    return;
  }
  /**
   * @return the text
   */
  public String getText()
  {
    return text;
  }
  /**
   * @return the hash
   */
  public byte[] getHash()
  {
    return hash;
  }
  /**
   * @return the hex
   */
  public String getHex()
  {
    return hex;
  }
  /**
   * Compare two IDs.
   * 
   * Currently we compare on the original string (for our purposes it should be the fastest).
   * 
   * @param id the ID to compare with
   * @return whether the two IDs are identical or not
   */
  public boolean equals( DNCViolationID id )
  {
    return( this.getText().equals(id.getText()) );
  }
  /**
   * Convert to the hex form.
   */
  @Override
  public String toString()
  {
    return( this.hex );
  }
}


src/kandalaya/dncviolation/LogItemComparator.java
/*
 * Compare log entries by timestamp for reporting DNC violations on Android.
 * 
 * Copyright (C) 2011, Raj Mathur <raju@kandalaya.org>
 * 
 * This file is part of DNCViolation.
 * 
 * DNCViolation is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DNCViolation is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with DNCViolation.  If not, see <http://www.gnu.org/licenses/>.
 */
// $Id: LogItemComparator.java 8 2011-09-18 05:30:54Z raju $

package kandalaya.dncviolation;

import java.util.Comparator;

	public class LogItemComparator implements Comparator<LogItem>
	{
	  //@Override
		public int compare( LogItem i1, LogItem i2 )
		{
		  long
		    t1 = i1.getTimestamp();
		  long
		    t2 = i2.getTimestamp();
		  if( t2 - t1 > 0 )
		  {
		    return( 1 );
		  }
		  else if( t1 == t2 )
		  {
		    return 0;
		  }
		  else
		  {
		    return( -1 );
		  }
		}
		public boolean equals(LogItem item1, LogItem item2)
		{
		  return( item1.getId().equals(item2.getId()) );
		}
	}
	


src/kandalaya/dncviolation/DncViolationSharedPreferences.java
/*
 * Handle preferences for reporting DNC violations on Android.
 * 
 * Copyright (C) 2011, Raj Mathur <raju@kandalaya.org>
 * 
 * This file is part of DNCViolation.
 * 
 * DNCViolation is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DNCViolation is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with DNCViolation.  If not, see <http://www.gnu.org/licenses/>.
 */
// $Id$

package kandalaya.dncviolation;

import android.content.Context;
import android.content.SharedPreferences;

public class DncViolationSharedPreferences
{
  /**
   * Name of preferences.
   */
  public final static String
    PREFERENCES_NAME = "DncViolationPreferences";
  /**
   * Get preference for deleting SMSes after reporting them.  true: delete after reporting; false: keep after reporting. 
   * @param context Context.
   * @return The user preference value.
   */
  public static boolean getDeleteSmsOnReport( Context context )
  {
    SharedPreferences
      preferences = context.getSharedPreferences(PREFERENCES_NAME, 0);
    boolean
      deleteSmsOnReport = preferences.getBoolean(context.getString(R.string.pref_key_delete_sms_on_report), false);
    return( deleteSmsOnReport );
  }
  /**
   * Get the number to report violations via SMS to.
   * @param context Context.
   * @return The number to report to.
   */
  public static String getReportingSmsNumber( Context context )
  {
    SharedPreferences
      preferences = context.getSharedPreferences(PREFERENCES_NAME, 0);
    String
      reportingSmsNumber = preferences.getString(context.getString(R.string.pref_key_reporting_sms_number), context.getString(R.string.default_reporting_sms_number));
    return( reportingSmsNumber );
  }
}


src/kandalaya/dncviolation/DncViolationPreferencesActivity.java
/*
 * User interface for editing preferences for reporting DNC violations on Android.
 * 
 * Copyright (C) 2011, Raj Mathur <raju@kandalaya.org>
 * 
 * This file is part of DNCViolation.
 * 
 * DNCViolation is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DNCViolation is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with DNCViolation.  If not, see <http://www.gnu.org/licenses/>.
 */
// $Id: DNCViolationActivity.java 8 2011-09-18 05:30:54Z raju $

package kandalaya.dncviolation;

import android.os.Bundle;
import android.preference.PreferenceActivity;

public class DncViolationPreferencesActivity extends PreferenceActivity
{
  @Override
  protected void onCreate( Bundle savedInstanceState )
  {
    super.onCreate(savedInstanceState);
    getPreferenceManager().setSharedPreferencesName(
        DncViolationSharedPreferences.PREFERENCES_NAME);
    addPreferencesFromResource(R.xml.preferences);
  }
}


src/kandalaya/dncviolation/DNCViolationDB.java
/*
 * Get call/SMS logs for reporting DNC violations on Android.
 * 
 * Copyright (C) 2011, Raj Mathur <raju@kandalaya.org>
 * 
 * This file is part of DNCViolation.
 * 
 * DNCViolation is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * DNCViolation is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with DNCViolation.  If not, see <http://www.gnu.org/licenses/>.
 */
// $Id: DNCViolationDB.java 8 2011-09-18 05:30:54Z raju $

package kandalaya.dncviolation;

import java.util.Collections;
import java.util.ArrayList;

import android.app.Activity;
import android.database.Cursor;
import android.util.Log;
import android.net.Uri;

import kandalaya.dncviolation.LogItem;
import kandalaya.dncviolation.LogItemComparator;

public class DNCViolationDB {
  /**
   * URI for the SMS inbox, needed to fetch SMSes.
   */
  static Uri
    smsInboxUri = Uri.parse("content://sms/inbox");
  /**
   * URI for SMS conversations, needed for deletion.
   */
  static Uri
    smsContentUri = Uri.parse("content://sms");
	/**
	 * Make a list of calls containing LogItems.
	 * 
	 * 	 @param activity The activity for which this instance was started (required for getContentProvider).
	 *   @param unknownOnly	Only include calls from numbers without cached names. 
	 *   @param nDays Only consider records newer than this many days.
	 */
	static ArrayList<LogItem>
	makeCallList(Activity activity, boolean unknownOnly, int nDays)
	{
	  Log.d("DNCViolation", "ndays = "+nDays);
		ArrayList<LogItem>
			mylist = new ArrayList<LogItem>();
		//
		// Set first day records from which we are interested in.
		//
    String
      strOrder;
    String
      dateFilter;
		//
		// Get call log
		//
		String[]
		       callStrFields = {
		    "_id",
				android.provider.CallLog.Calls.NUMBER, 
				android.provider.CallLog.Calls.CACHED_NAME,
				android.provider.CallLog.Calls.DATE,
		};
		String
			callerNameFilter = "";
		strOrder = android.provider.CallLog.Calls.DATE + " DESC";
		dateFilter = " and julianday('now') - julianday(" + android.provider.CallLog.Calls.DATE + "/1000, 'unixepoch') <= " + nDays;
		Log.d("DNCViolation", "date filter = <"+dateFilter+">");
		if( unknownOnly )
		{
			callerNameFilter = " and " + android.provider.CallLog.Calls.CACHED_NAME + " is null";
			Log.d("DNCViolation", "Set filter to null");
		}
		//Log.d("DNCViolation", "order: "+strOrder+", filter: "+callerNameFilter);
		Log.d("DNCViolation", "start query");
		Cursor
			callCursor = activity.getContentResolver().query(
					android.provider.CallLog.Calls.CONTENT_URI,
					callStrFields,
					android.provider.CallLog.Calls.TYPE + " = " + android.provider.CallLog.Calls.INCOMING_TYPE
						+ callerNameFilter
						+ dateFilter
						,
					null,
					null
			);
		// get start of cursor
		Log.d("DNCViolation", "start query");
		if( callCursor.moveToFirst() )
		{
			int
				i = 0;
			// loop through cursor 
			do
			{
			  int
			    id = callCursor.getInt(0);
				String
					number = callCursor.getString(1);
				String
					name = callCursor.getString(2);
				long
					timestamp = callCursor.getLong(3);
				LogItem
					item = new LogItem();
				item.setNumber(number);
				item.setName(name);
				item.setTimestamp(timestamp);
				//Log.d("DNCViolation", "timestamp " + timestamp);
				item.setSelected(false);
				item.setCall_type('C');
        item.setId("C", id);
				// Log.d("DNCViolation", number+", "+name+", "+date.toLocaleString());
				mylist.add(item);
				i++;
			} while (callCursor.moveToNext());
			Log.d("DNCViolation", "got " + i + " call records");
		}
		Log.d("DNCViolation", "done call query");
		//
		// Get SMS log
		//
		/*// What are the fields anyway?
		Cursor mCursor = activity.getContentResolver().query(smsUri, null, null, null, null);

		StringBuffer info = new StringBuffer();
		for( int i = 0; i < mCursor.getColumnCount(); i++) {
		    info.append("Column: " + mCursor.getColumnName(i) + "\n");
		}
		Log.d("DNCViolation", new String(info));*/
		//
		// Columns:
		// _id, thread_id, address, person, date, protocol, read, status,
		// type, reply_path_present, subject, body, service_center, locked,
		// error_core, seen, deletable, hidden, group_id, group_type,
		// delivery_date

		//
		// Get the SMSes
		String[]
		       smsStrFields =
		       {
		            "_id",
                "date",
                "address",
                "subject",
                "protocol",
                "body",
                "person",
		       };
    strOrder = "date DESC";
    dateFilter = "julianday('now') - julianday(date/1000, 'unixepoch') <= " + nDays;
    Log.d("DNCViolation", "SMS date filter = <"+dateFilter+">");
		Log.d("DNCViolation", "start SMS query");
		Cursor
			smsCursor = activity.getContentResolver().query(
					smsInboxUri,
					smsStrFields,
					dateFilter,
					null,
					strOrder
			);
		// get start of cursor
		Log.d("DNCViolation", "start SMS query");
		if( smsCursor.moveToFirst() )
		{
			int
				i = 0;
			// loop through cursor 
			do
			{
			  int
			    id = smsCursor.getInt(0);
				long
					timestamp = smsCursor.getLong(1);
				String
					address = smsCursor.getString(2);
				String
					subject = smsCursor.getString(3);
				String
					body = smsCursor.getString(5);
				LogItem
					item = new LogItem();
				item.setNumber(address);
				item.setName(address);
				item.setTimestamp(timestamp);
				item.setSelected(false);
				item.setCall_type('M');
				// Get extract of SMS to show in name field
				String
					name = new String("");
				if( subject != null )
				{
					name += subject + " | "; 
				}
				name += body;
				item.setNumber(address);
				item.setName(name);
				item.setCall_type('S');
				item.setSelected(false);
				item.setId("S", id);
				//Log.d("DNCViolation", address+", "+person+", "+protocol+", "+subject+", "+item.getName()+", "+date.toLocaleString());
				mylist.add(item);
				i++;
			} while (smsCursor.moveToNext());
			Log.d("DNCViolation", "got " + i + " SMS records");
		}
		Log.d("DNCViolation", "done SMS query");
		Collections.sort(mylist, new LogItemComparator());
		for( int i = 0; i < mylist.size(); i++ )
		{
		  mylist.get(i).setIndex(i);
		}
		return( mylist );
	}
	/**
	 * Delete the SMS given the internal Android ID.  Code copied from
	 * http://stackoverflow.com/questions/419184/how-to-delete-an-sms-from-the-inbox-in-android-programmatically
	 * 
	 * @param activity Activity (needed for getContentResolver).
	 * @param id The internal Android ID of the SMS.
	 * @return true on success, false on failure to delete.
	 */
	public static boolean deleteSMS(Activity activity, int id)
	{
	  //
	  // Get the thread ID
	  //
	  int
	    threadId;
	  String[]
	         smsStrFields =
	         {
	            "_id",
	            "thread_id"
	         };
    Log.d("DNCViolation", "start SMS thread query");
    Cursor
      smsCursor = activity.getContentResolver().query(
          smsInboxUri,
          smsStrFields,
          "_id = ?",
          new String[] {String.valueOf(id)},
          null
      );
    // get start of cursor
    Log.d("DNCViolation", "start SMS query");
    if( smsCursor.moveToFirst() )
    {
      threadId = smsCursor.getInt(1);
      Log.d("DNCViolation", "got SMS thread id "+threadId);
    }
    else
    {
      Log.d("DNCViolation", "Cannot find SMS with ID "+id);
      return( false );
    }
    //
    // Try to delete the specific SMS
    //
    Log.d("DNCViolation", "deleting SMS thread id "+threadId+" _id "+id);
    int
      deleted;
    deleted = activity.getContentResolver().delete(smsContentUri, "thread_id = ? and _id = ?", new String[] {String.valueOf(threadId), String.valueOf(id)});
    if( deleted > 0 )
    {
      Log.d("DNCViolation", deleted+" SMS deleted");
    }
    else
    {
      Log.d("DNCViolation", "Couldn't delete the SMS");
    }
    return( deleted > 0 );
	}
  /**
   * Delete the SMS given ID in our format.  Extract "NNNNN" from the "{S/C}-NNNNN" format
   * and pass to deleteSMS(Activity, int).
   * 
   * @param activity Activity (needed for getContentResolver).
   * @param id Our ID of the SMS.
   * @return true on success, false on failure to delete.
   */
	public static boolean deleteSMS( Activity activity, String id )
	{
	  String []
	    splitted = id.split("-");
	  String
	    idPart = splitted[1];
	  Log.d("DNCViolation", "Calling deleteSMS("+idPart+")");
	  
	  int
	    numericId = Integer.parseInt(idPart);
	  return( deleteSMS(activity, numericId) );
	}
}
